File: vehicle/haskell/src/Car.hs
module Car (Car (..)) where

import Vehicle (Vehicle)

class Vehicle carType => Car carType where
  getEngine :: carType -> String
  getEngine _ = "four-stroke engine"

  drive :: carType -> IO ()
  drive car = do
    putStrLn ("Car engine: " <> getEngine car)
    putStrLn "#############################"

  simulate :: carType -> IO ()
  simulate = drive

File: vehicle/rust/src/car.rs
use crate::Vehicle;

pub trait Car: Vehicle {
  fn get_engine(&self) -> String {
    String::from("four-stroke engine")
  }

  fn drive(&self) {
    println!("Car engine: {}", Car::get_engine(self));
    println!("#############################");
  }

  fn simulate(&self) {
    self.drive();
  }
}
